### DISCUSSION TITLE
Does Arduino "Serial" work via SWD?

### DISCUSSION THREAD

**Discussion by olikraus** [stm32duino/Arduino_Core_STM32#1647] [Q&A]
Might be a stupid question, but maybe it can be confirmed.
I have a STM32G0 device connected via SWD.
Will the "Serial.print" work via SWD interface (somehow) or do I need an additional USB/USART interface?

I assume a second USB-USART converter is required for this, right?

So, if I want to see some Serial.print output in Arduino IDE Serial Monitor I need both: SWD ST-Link adapter and USB-USART converter (occupying two USB slots), correct?

Thanks,
Oliver

**Comment by fpistm** [comment]
> Will the "Serial.print" work via SWD interface (somehow) or do I need an additional USB/USART interface?

It is not managed with this core. It will required to enable the SWV and rewrite the `_write()` syscall function to use `ITM_SendChar((*ptr++));` and it is mainly used when you debug thanks SWD interface.

**Reply by olikraus** [reply]
ok, so I use STM32duino out of the box with SWD ST-Link and if I want to use Serial Monitor, then I need an extra USB-USART convert becuase Serial.print goes to USART only, correct?

**Comment by ABOSTM** [comment]
Hi @olikraus,

> Will the "Serial.print" work via SWD interface (somehow) or do I need an additional USB/USART interface?

Strictly speaking it is not via SWD, nevertheless you don't need extra USB/USART converter, :
In all STM32 board equipped with ST-Link (like Nucleo boards, or Disco boards, ...), an USART instance is connected from MCU to ST-Link sometime called Virtual Com Port (VCP). And ST-Link makes the rest to create a COM port on your PC. On those board, Stm32duino is generally configured to map the generic "Serial" to this instance of USART, so that Serial.print will work via ST-Link.
Example: on STM32G0316-DISCO you can see on schematics the link between ST-link and MCU
https://www.st.com/resource/en/schematic_pack/mb1454-g031j6-b02_schematic.pdf

In other words, I never use external USB/USART converter with STM32 chip,  but only the ST-link.

**Reply by olikraus** [reply]
Wow, amazing, thanks for the long answer. Still an question is there: How do I need to cofigure the Arduino Serial object to send to this Virtual Com Port? I mean: Do I need to do some special setup for the Serial object im my .ino file?

**Reply by ABOSTM** [reply]
You just need to define the default `Serial `instance with the right pin to be used 
https://github.com/stm32duino/Arduino_Core_STM32/blob/de25178ca432912badd135445f80baee198425d0/variants/STM32G0xx/G031J(4-6)M_G041J6M/variant_generic.h#L102

  #define SERIAL_UART_INSTANCE  xxx
   ...
  #define PIN_SERIAL_RX         xxx
   ...
  #define PIN_SERIAL_TX         xxx

This will simply map default "Serial" object variable to a specific USART instance with it associated pins.
Then, as you have a custom board, you need to connect those pins to STLink. (warning you need to cross pins: TX from MCU is connected to RX from ST-link)

**Reply by olikraus** [reply]
Ok, if understand the pinout of the st-link correctly, then there are two more pins for ST-Link: RX and TX (like described here: https://stackoverflow.com/questions/66470107/what-is-use-of-rx-and-tx-in-st-link-v2).
So on my own device, I have to connect SWIO, SWCLK, RX and TX (four different pins). ST-Link will then act as a USART - USB converter (virtual COM port), correct? 

I also assume, that it is not possible to use only two pins, like sharing for example SWIO and RX one pin and SWCLK and TX another pin, correct?

**Reply by ABOSTM** [reply]
To go through SWIO, refer to @fpistm comment:

> It is not managed with this core. It will required to enable the SWV and rewrite the _write() syscall function to use ITM_SendChar((*ptr++)); and it is mainly used when you debug thanks SWD interface.

**Reply by olikraus** [reply]
ok, thanks for all the answers and clarification.
